### ISSUE TITLE
Dropout in diffrax.diffeqsolve

### ISSUE THREAD

**Issue by stefangroha** [patrick-kidger/diffrax#105]
Hello,

I am relatively new to jax and equinox/diffrax, so sorry if this is trivial.

I wanted to add some randomness to a neural ODE. Due to the functional form of the vector field I do not want to use Brownian motion as a control term, but I would like to include dropout in the network that parametrizes the vector field. 

I want to sample a different dropout for every time step, which requires a new key at every step, however it is not clear to me how I would achieve this using diffrax.diffeqsolve. I could step through the solver and split the key at every step, but I was wondering if there is a way while still using diffrax.diffeqsolve. 

Thank you for the really great package!

**Comment by patrick-kidger** [comment]
Hey there! So adding randomness is a pretty odd thing to do, from the point of view of a differential equation solver. (I recognise the desire to add it from an ML perspective though.)

The reason is that what you've got is no longer an ODE: it's no longer deterministic. This is mostly a consideration when adaptive step sizing. Normally: each time the solver makes a step, the step size controller checks that the numerical error isn't too large. If the error is too large, the step will be rejected and a smaller step tried instead. But in the presence of randomness: you might get a step rejection just because of the random sampling producing a large-error step, and then the retried step gets accepted just because you used different randomness and got a different small-error vector field. In doing so you don't actually get the randomness you were expecting: there is a bias towards whatever part of the random sample space produces "easy to solve" vector fields.

In other words, "neural ODE w/ dropout w/ adaptive solver" actually isn't a mathematically well-formed idea.

Despite which: yes, this is absolutely possible! You should probably use a fixed (non-adaptive) solver -- i.e. `diffeqsolve(..., stepsize_controller=ConstantStepSize())`, don't use `stpesize_controller=PIDController()` -- and it's just worth recognising that what you're doing is mathematically a bit questionable ;)

(I have used dropout in neural ODEs before quite successfully.)

In terms of the code, the appropriate behaviour is to use `jax.random.fold_in` to update a key based on the value of time. (Just bitcast the floating-point value of time to an integer.) For example:
```
from diffrax import diffeqsolve
from diffrax.misc import force_bitcast_convert_type
import jax.numpy as jnp
import jax.random as jr

def vector_field(t, y, args):
    key = args
    t_ = force_bitcast_convert_type(t, jnp.int32)
    key = jr.fold_in(key, t_)
    ...  # use key

key = jr.PRNGKey(0)
diffeqsolve(vector_field, ..., args=key)
```
or a larger example if you're using this as part of an Equinox model:
```
from diffrax import diffeqsolve
from diffrax.misc import force_bitcast_convert_type
import equinox as eqx
import jax.numpy as jnp
import jax.random as jr

class VectorField(eqx.Module):
    dropout: eqx.nn.Dropout
    key: jr.PRNGKey

    def __init__(self, key):
        dkey, kkey = jr.split(key)
        self.dropout = eqx.nn.Dropout(key=dkey)
        self.key = kkey

    def __call__(self, t, y, args):
        t_ = force_bitcast_convert_type(t, jnp.int32)
        key = jr.fold_in(self.key, t_)
        return self.dropout(y, key=key)

key = jr.PRNGKey(0)
vector_field = VectorField(key)
diffeqsolve(vector_field, ...)
```

Note that `force_bitcast_convert_type` is technically an internal API for Diffrax. I'd recommend copy-pasting the source code for `force_bitcast_convert_type` into your application if you want to be certain of stability. (Realistically I don't think `force_bitcast_convert_type` is going anywhere though.)

Disclaimer, I've not tested any of the above code samples.

**Comment by stefangroha** [comment]
Hey, thank you so much for your answer! 

This works for the solution I want it to be. I do realize that it is pretty questionable mathematically and comes from an ML view, but it was suggested to me just to improve robustness of the solution.

Thank you again!
